tmp/trycabal.hs

```diff
@@ -0,0 +1,49 @@
+{-# LANGUAGE DeriveDataTypeable  #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+import Data.Data
+import Distribution.PackageDescription.Parsec
+import Distribution.Types.BuildInfo
+import Distribution.Types.CondTree
+import Distribution.Types.GenericPackageDescription
+import Distribution.Types.Library
+import Language.Haskell.Extension
+import Text.Pretty.Simple
+import qualified Data.ByteString as BS
+
+import Data.Generics.Uniplate.Data as U
+
+
+pp :: (Show a) => a -> IO ()
+pp = pPrintOpt NoCheckColorTty defaultOutputOptionsDarkBg
+
+
+data X = A [Extension]
+       | B Extension
+       | C Int
+       | D X X X
+       deriving (Show, Data, Typeable)
+
+
+main = do
+    let struct = D (B (EnableExtension OverlappingInstances))
+                   (A [EnableExtension DoRec, DisableExtension RecursiveDo, UnknownExtension "ZZZ"])
+                   (C 42)
+        extensions = [x | EnableExtension x <- U.universe struct]
```

Comment:
Solution is simple: use universeBi :: _ => from -> [to] instead of uniplate :: _ => to -> [to].